std/process.d

```diff
@@ -450,94 +465,133 @@ private Pid spawnProcessImpl(in char[][] args,
+                rlimit r;
+                if (getrlimit(RLIMIT_NOFILE, &r) != 0)
+                {
+                    abortOnError(forkPipeOut, InternalError.getrlimit, .errno);
+                }
+                immutable maxDescriptors = cast(int) r.rlim_cur;
+
+                // The above, less stdin, stdout, and stderr
+                immutable maxToClose = maxDescriptors - 3;
+
+                // Call poll() to see which ones are actually open:
+                pollfd* pfds = cast(pollfd*) malloc(pollfd.sizeof * maxToClose);
+                foreach (i; 0 .. maxToClose)
+                {
+                    pfds[i].fd = i + 3;
+                    pfds[i].events = 0;
+                    pfds[i].revents = 0;
+                }
+                if (poll(pfds, maxToClose, 0) >= 0)
+                {
+                    foreach (i; 0 .. maxToClose)
+                    {
+                        // don't close pipe write end
+                        if (pfds[i].fd == forkPipeOut) continue;
+                        // POLLNVAL will be set if the file descriptor is invalid.
+                        if (!(pfds[i].revents & POLLNVAL)) close(pfds[i].fd);
+                    }
+                }
+                else
+                {
+                    // Fall back to closing everything.
+                    foreach (i; 3 .. maxDescriptors)
+                    {
+                        if (i == forkPipeOut) continue;
+                        close(i);
+                    }
+                }
+            }
+            else // This is already done if we don't inherit descriptors.
             {
-                abortOnError(forkPipeOut, InternalError.getrlimit, .errno);
+                // Close the old file descriptors, unless they are
+                // either of the standard streams.
+                if (stdinFD  > STDERR_FILENO)  close(stdinFD);
+                if (stdoutFD > STDERR_FILENO)  close(stdoutFD);
+                if (stderrFD > STDERR_FILENO)  close(stderrFD);
             }
-            immutable maxDescriptors = cast(int) r.rlim_cur;
 
-            // The above, less stdin, stdout, and stderr
-            immutable maxToClose = maxDescriptors - 3;
+            // Execute program.
+            core.sys.posix.unistd.execve(argz[0], argz.ptr, envz);
 
-            // Call poll() to see which ones are actually open:
-            pollfd* pfds = cast(pollfd*) malloc(pollfd.sizeof * maxToClose);
-            foreach (i; 0 .. maxToClose)
+            // If execution fails, exit as quickly as possible.
+            abortOnError(forkPipeOut, InternalError.exec, .errno);
+        }
+
+        if (config & Config.detached)
+        {
+            auto secondFork = core.sys.posix.unistd.fork();
+            if (secondFork == 0)
             {
-                pfds[i].fd = i + 3;
-                pfds[i].events = 0;
-                pfds[i].revents = 0;
+                setsid();
```

Comment:
Oh, well, not all apps belong to the same session actually. Only those that was ran from desktop. If it runs from terminal, session is different.
The point is that user may want to have it in another session or in the same. I think it's better to leave to user to decide.